### ISSUE TITLE
Compatibility renderer outperforms Forward+ renderer (`GPUParticles3D`)

### ISSUE THREAD

**Issue by TheYellowArchitect** [godotengine/godot#97903]
### Tested versions

- Reproducible in 4.4 dev3, dev2 (and probably previous versions)

### System information

Godot v4.4.dev3 - Artix Linux #1 SMP PREEMPT_DYNAMIC Wed, 02 Oct 2024 15:03:06 +0000 on Tty - X11 display driver, Multi-window, 1 monitor - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1050 Ti (nvidia; 560.35.03) - AMD Ryzen 5 2600 Six-Core Processor (12 threads)

### Issue description

I was making a smoke grenade, so I went to use `GPUParticles3D`. I basically made quad meshes (backs are culled) with smoke textures, and with only **80** particles on-screen, the FPS tanks - a single `GPUParticles3D`!
So if a few smoke grenades are thrown, the game becomes unplayable.

Yet, this does not happen at the compatibility renderer. I made a MRP and benchmarked them with 3 trials each.

Single Smoke Grenade `GPUParticles3D`:
Renderer | Forward+ | Compatibility |
--- | --- | --- |
Lowest FPS | 63 | 201 |
Lowest FPS | 73 | 206 |
Lowest FPS | 74 | 218 |

2 Smoke Grenades `GPUParticles3D`:
Renderer | Forward+ | Compatibility |
--- | --- | --- |
Lowest FPS | 40 | 151 |
Lowest FPS | 52 | 176 |
Lowest FPS | 57 | 170 |

![smoke-grenade-54-fps-forward-renderer](https://github.com/user-attachments/assets/1ed79312-1208-4c94-beb0-60d51b8fb97c)

I think this issue is important as it addresses the "_godot sucks at 3D, its performance is abysmal._" complaint from users.

### Steps to reproduce

Make any `GPUParticles3D` with quad mesh and make a practical effect, so it uses alpha/transparency.

### Minimal reproduction project (MRP)

[smoke-grenade-room.zip](https://github.com/user-attachments/files/17270975/smoke-grenade-room.zip)

Use arrow keys to zoom up (and WS if you want to move front/back)

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot/issues/68959.

This is expected, the Compatibility rendering method has a lower base cost for simple scenes, and also a lower fill rate cost since its shaders are *much* simpler.

The use case shown in the screenshot is also a pathological use case for any renderer. Having this many large overlapping planes in front of the camera imposes an extreme fillrate cost on the GPU. Here, the renderer with the simplest shaders will always win. Also, remember to always use the **Unshaded** shading mode for large particles in front of the screen, and consider using *more* particles that are slower individually to reduce fillrate requirements. Here, using more particles can actually be *faster* despite the higher polygon count.

The Forward+ and Mobile rendering methods have much more advanced features, which generally implies a higher base cost (i.e. simple scenes are slower to render). However, in return, complex scenes with lots of geometry and draw calls are generally able to scale better to these rendering methods. Choose the right tool for the job 🙂

**Comment by TheYellowArchitect** [comment]
I thank you for the swift and detailed answer. I had no idea about the differences between the renderers, I thought Forward+ was objectively faster than compatibility. Nor did I think it was possible for more particles with less scale, to be more performant (I confirmed, it is true :+1: )

>Also, remember to always use the Unshaded shading mode

Did a quick benchmark:
2 Smoke Grenades `GPUParticles3D`:
Renderer | Forward+ | Compatibility |
--- | --- | --- |
Lowest FPS | 167 | 240 (max) |

>If you need the particles to be shaded, use [per-vertex shading](https://github.com/godotengine/godot/pull/83360) which is significantly faster than per-pixel shading in this situation.

I built that specific commit, here are the benchmarks (2 trials) for per-vertex shading, its a good improvement

2 Smoke Grenades `GPUParticles3D`:
Renderer | Forward+ | Compatibility |
--- | --- | --- |
Lowest FPS | 85 | 229 |
Lowest FPS | 84 | 216 |
Lowest FPS | 93 | 238|

>In Forward+, you may also get better performance from enabling volumetric fog and using FogVolume nodes to achieve smoke effects.

It probably isn't ideal for my game, as the camera is top-down. Also I want it to work like a real smoke grenade, where it pours smoke fast to fill a room, instead of the "packed smoke" approach of counter strike (and all other FPS which copy-paste its smoke grenade)
I will try it at some point and confirm.

For now, I will tweak the per-vertex smoke grenade in hopes I make it look as intended, with more particles and less overlaps.

>The Forward+ and Mobile rendering methods have much more advanced features

Is there a list of them? Decals are the ones I have noticed, but what else? The only documentation I found on their differences is this: https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/internal_rendering_architecture.html
